ouroboros-consensus/test-consensus/Test/Dynamic/Util.hs

```diff
@@ -251,3 +260,65 @@ longestCrowdedRun (LeaderSchedule m) = fst
         | n == 0    = (x, y)
         | n == 1    = (x, noRun)
         | otherwise = let y' = incCrowdedRun slot y in (max x y', y')
+
+{-------------------------------------------------------------------------------
+  Node Connection Schedule - when each node joins the test network
+-------------------------------------------------------------------------------}
+
+-- | A list of durations as number of slots.
+-- Each duration is the amount of time after a node joins the network before
+-- either the next node does or the test ends.
+--
+-- ASSUMPTIONS:
+-- * Nodes join in the order of their Id.
+-- * Node 0 joins first, at slot 0, so the first duration in the list occurs after that.
+-- * Nodes other than Node 0 join after some possibly empty duration
+--   since the previous node joined.
+-- * The last duration in the list is the run duration after the last node joins.
+-- * The number of durations is therefore the number of nodes.
+newtype ConnectionPlan = ConnectionPlan [Int]
```

Comment:
Nice and simple :slightly_smiling_face: 

However, I'm a bit worried that there's too much (documented, yet) non-obvious information captured in this `[Int]`:
* The first node always starts at 0, so it is absent from the list.
* The values are "accumulating slots", i.e. to know at which slot the *n*th node joins, you have to add the first *n* elements in the list. When a test fails with a very long list, figuring out at which slot the last node join will be not be immediately evident (or we should mention the output `interpretConnectionPlan` using `counterexample`).
* The last value in the list is not how long to wait before the last node joins, but how long to run the complete network after the last node has joined.

I think this type is easy to generate and shrink, but it may not be so human-friendly :slightly_smiling_face: 

Also, will it be easy to extend this type to account for nodes disconnecting or network partitions? (I know you can't always design everything with all future requirements in mind).

A proposal:

```haskell
data ConnectionPlan = ConnectionPlan
  { nodesJoinAt   :: NonEmpty SlotNo
  , networkEndsAt :: SlotNo
  }
```
* `NonEmpty`: because there must always be at least one node
* `NonEmpty SlotNo`: you can directly see which node joins at which number (even easier in a `Map`, but other invariants will then be more work to maintain). They don't have to join in order, so the list doesn't have to be monotonic.
* Invariant: `networkEndsAt > maximum nodesJoinAt`

This is just a proposal, of course. What are your thoughts?

Reply:
Just reread a comment of yours: indeed, we don't care about the order, which is an argument for "accumulating slots".

Reply:
Weeks laters, this is resolved as he requested.